integrations/bytewax/orderbook.py

```diff
@@ -0,0 +1,206 @@
+            for update in data["changes"]:
+                price = float(update[1])
+                size = float(update[2])
+            if update[0] == "sell":
+                # first check if the size is zero and needs to be removed
+                if size == 0.0:
+                    try:
+                        del self.asks[price]
+                        # if it was the ask price removed,
+                        # update with new ask price
+                        if price <= self.ask_price:
+                            self.ask_price = min(self.asks.keys())
+                    except KeyError:
+                        # don't need to add price with size zero
+                        pass
+                else:
+                    self.asks[price] = size
+                    if price < self.ask_price:
+                        self.ask_price = price
+            if update[0] == "buy":
+                # first check if the size is zero and needs to be removed
+                if size == 0.0:
+                    try:
+                        del self.bids[price]
+                        # if it was the bid price removed,
+                        # update with new bid price
+                        if price >= self.bid_price:
+                            self.bid_price = max(self.bids.keys())
+                    except KeyError:
+                        # don't need to add price with size zero
+                        pass
+                else:
+                    self.bids[price] = size
+                    if price > self.bid_price:
+                        self.bid_price = price
+        return self, {
+            "ticker": data["product_id"],
+            "timestamp": datetime.strptime("2023-07-17T18:03:25.947610Z", "%Y-%m-%dT%H:%M:%S.%fZ").replace(
+                tzinfo=timezone.utc),
+            "bid": self.bid_price,
+            "bid_size": self.bids[self.bid_price],
+            "ask": self.ask_price,
+            "ask_price": self.asks[self.ask_price],
+            "spread": self.ask_price - self.bid_price,
+        }
+
+
+def get_flow(feature_group_name, feature_group_version):
+    flow = Dataflow()
+    flow.input("input", CoinbaseFeedInput(["BTC-USD", "ETH-USD", "BTC-EUR", "ETH-EUR"]))
+    flow.map(json.loads)
+    # {
+    #     'type': 'l2update',
+    #     'product_id': 'BTC-USD',
+    #     'changes': [['buy', '36905.39', '0.00334873']],
+    #     'time': '2022-05-05T17:25:09.072519Z',
+    # }
+    flow.map(key_on_product)
+    # ('BTC-USD', {
+    #     'type': 'l2update',
+    #     'product_id': 'BTC-USD',
+    #     'changes': [['buy', '36905.39', '0.00334873']],
+    #     'time': '2022-05-05T17:25:09.072519Z',
+    # })
+    flow.stateful_map("order_book", lambda: OrderBook(), OrderBook.update)
+    # ('BTC-USD', (36905.39, 0.00334873, 36905.4, 1.6e-05, 0.010000000002037268))
+    flow.filter(
+        lambda x: x[-1]["spread"] / x[-1]["ask"] > 0.0001
+    )  # filter on 0.1% spread as a per
```

Comment:
I might remove this, and let the downstream model decide on a threshold. It will significantly increase the volume, but it feels like the right pattern.